### PR TITLE
Optimize getbbox() and getextrema() routines

### PR DESCRIPTION
**Pull request by homm**
Code for ipython:

```python
from PIL import Image
im = Image.new('L', (2048, 2048), 0)
print('Worst case:', im.getbbox())
%timeit im.getbbox()
im = Image.open('tests/images/face1.png').resize((2048, 2048))
print('Normal case:', im.getbbox())
%timeit im.getbbox()
im = Image.open('tests/images/dices.png').resize((2048, 2048))
print('Normal case:', im.getbbox())
%timeit im.getbbox()
im = im.convert('RGB').convert('RGBA')
print('Best case:', im.getbbox())
%timeit im.getbbox()
im = im.convert('L')
print('Extrema:', im.getextrema())
%timeit im.getextrema()
```

## `main` branch

```python
Worst case: None
2.31 ms ± 3.29 μs per loop (mean ± std. dev. of 7 runs, 100 loops each)
Normal case: (150, 474, 1802, 2048)
3.97 ms ± 519 ns per loop (mean ± std. dev. of 7 runs, 100 loops each)
Normal case: (286, 70, 1724, 1797)
3.86 ms ± 461 ns per loop (mean ± std. dev. of 7 runs, 100 loops each)
Best case: (0, 0, 2048, 2048)
4.6 ms ± 464 ns per loop (mean ± std. dev. of 7 runs, 100 loops each)
Extrema: (0, 255)
2.3 ms ± 468 ns per loop (mean ± std. dev. of 7 runs, 100 loops each)
```

## `optimize-getbbox` branch

```python
Worst case: None
1.16 ms ± 1.65 μs per loop (mean ± std. dev. of 7 runs, 1,000 loops each)
Normal case: (150, 474, 1802, 2048)
722 μs ± 161 ns per loop (mean ± std. dev. of 7 runs, 1,000 loops each)
Normal case: (286, 70, 1724, 1797)
874 μs ± 260 ns per loop (mean ± std. dev. of 7 runs, 1,000 loops each)
Best case: (0, 0, 2048, 2048)
2.51 μs ± 18.1 ns per loop (mean ± std. dev. of 7 runs, 100,000 loops each)
Extrema: (0, 255)
97.6 μs ± 5.74 ns per loop (mean ± std. dev. of 7 runs, 10,000 loops each)
```

```
Worst case: 1.99x speedup
Normal case: 5.5x speedup
Normal case: 4.4x speedup
Best case: 1833x speedup
Extrema: 23.5x speedup
```

## `dices.png` test image

dices

## `face1.png` test image

face1

### CODE REVIEW

File: src/libImaging/GetBBox.c [hunk 1]
```diff
@@ -30,26 +30,47 @@ ImagingGetBBox(Imaging im, int bbox[4], int alpha_only) {
...
+    /* Check that we got a box */                            \
+    if (bbox[1] < 0)                                         \
+        return 0; /* no data */                              \
+    /* second stage: looking for any px from bottom */       \
+    for (y = im->ysize - 1; y >= bbox[1]; y--) {             \
+        has_data = 0;                                        \
+        for (x = 0; x < im->xsize; x++)                      \
+            if (im->image[y][x] & mask) {                    \
+                has_data = 1;                                \
+                bbox[3] = y + 1;                             \
```
Comment (Yay295) [inline src/libImaging/GetBBox.c:56]: I think `bbox[2]` could be set here.

Comment (homm) [reply src/libImaging/GetBBox.c:56]: No, since x is iterated from left to right, which is faster than opposite.

Comment (Yay295) [reply src/libImaging/GetBBox.c:56]: oh, right. It could be `bbox[0] = min(x,bbox[0])` then, no?

Comment (homm) [reply src/libImaging/GetBBox.c:56]: This makes sense! thanks

File: src/libImaging/GetBBox.c [hunk 2]
```diff
@@ -30,26 +30,57 @@ ImagingGetBBox(Imaging im, int bbox[4], int alpha_only) {
...
+    }                                                        \
+    /* third stage: looking for left and right boundaries */ \
+    for (y = bbox[1]; y < bbox[3]; y++) {                    \
+        for (x = 0; x < bbox[0]; x++) {                      \
+            if (im->image[y][x] & mask) {                    \
+                bbox[0] = x;                                 \
+                break;                                       \
+            }                                                \
+        }                                                    \
+        for (x = im->xsize - 1; x >= bbox[2]; x--) {         \
```
Comment (nulano) [inline src/libImaging/GetBBox.c:78]: Have you considered iterating left to right here as well?

Comment (homm) [reply src/libImaging/GetBBox.c:78]: It's very weird, but changing direction in this place leads to degradation on all three platforms (AWS c7i, c7a, c7g instances). My guess is GCC uses some optimization when there is break inside condition in a loop.

**Comment by radarhere** [comment]
Could you provide a copy of face1.png?

**Comment by homm** [comment]
@radarhere Ok, added, but it's just some random image from stocks)

**Comment by homm** [comment]
Rebased

**Pull request merged**